s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/SourceRecordIterator.java

```diff
@@ -122,50 +118,48 @@ private Iterator<List<ConsumerRecord<byte[], byte[]>>> createIteratorForCurrentF
                     bucketName);
 
-            return getObjectIterator(inputStream, finalTopic, defaultPartitionId, defaultStartOffsetId, outputWriter,
+            return getObjectIterator(inputStream, finalTopic, defaultPartitionId, defaultStartOffsetId, transformer,
                     partitionMap);
         }
     }
 
     @SuppressWarnings("PMD.CognitiveComplexity")
-    private Iterator<List<ConsumerRecord<byte[], byte[]>>> getObjectIterator(final InputStream valueInputStream,
-            final String topic, final int topicPartition, final long startOffset, final OutputWriter outputWriter,
+    private Iterator<AivenS3SourceRecord> getObjectIterator(final InputStream valueInputStream, final String topic,
+            final int topicPartition, final long startOffset, final Transformer transformer,
             final Map<String, Object> partitionMap) {
         return new Iterator<>() {
-            private List<ConsumerRecord<byte[], byte[]>> nextRecord = readNext();
-            private List<ConsumerRecord<byte[], byte[]>> readNext() {
+            private final Iterator<AivenS3SourceRecord> internalIterator = readNext().iterator();
 
-                final Optional<byte[]> optionalKeyBytes = Optional.ofNullable(currentObjectKey)
-                        .map(k -> k.getBytes(StandardCharsets.UTF_8));
-                final List<ConsumerRecord<byte[], byte[]>> consumerRecordList = new ArrayList<>();
+            private List<AivenS3SourceRecord> readNext() {
+                final byte[] keyBytes = currentObjectKey.getBytes(StandardCharsets.UTF_8);
+                final List<AivenS3SourceRecord> sourceRecords = new ArrayList<>();
 
                 int numOfProcessedRecs = 1;
                 boolean checkOffsetMap = true;
-                for (final Object record : outputWriter.getRecords(valueInputStream, topic, topicPartition)) {
-                    final byte[] valueBytes = outputWriter.getValueBytes(record, topic, s3SourceConfig);
-                    if (offsetManager.getOffsets().containsKey(partitionMap) && checkOffsetMap) {
-                        final Map<String, Object> offsetVal = offsetManager.getOffsets().get(partitionMap);
-                        if (offsetVal.containsKey(offsetManager.getObjectMapKey(currentObjectKey))) {
-                            final long offsetValue = (long) offsetVal
-                                    .get(offsetManager.getObjectMapKey(currentObjectKey));
-                            if (numOfProcessedRecs <= offsetValue) {
-                                numOfProcessedRecs++;
-                                continue;
-                            }
-                        }
+                for (final Object record : transformer.getRecords(valueInputStream, topic, topicPartition,
+                        s3SourceConfig)) {
+                    if (offsetManager.shouldSkipRecord(partitionMap, currentObjectKey, numOfProcessedRecs)
```

Comment:
should not we skip already processed records while reading/transforming and not when we did that already? Such way we will not have to keep already processed records in memory. Ideally if possible would be good to start reading from a specific offset in a file even

Reply:
This is an area of improvement. Ticket. : https://aiven.atlassian.net/browse/KCON-36